src/js/tools/sentryMetroSerializer.ts

```diff
@@ -0,0 +1,256 @@
+import * as crypto from 'crypto';
+import type { MixedOutput, Module, ReadOnlyGraph, SerializerOptions } from 'metro';
+import type { SerializerConfigT } from 'metro-config';
+import * as baseJSBundle from 'metro/src/DeltaBundler/Serializers/baseJSBundle';
+import * as sourceMapString from 'metro/src/DeltaBundler/Serializers/sourceMapString';
+import * as bundleToString from 'metro/src/lib/bundleToString';
+import CountingSet from 'metro/src/lib/CountingSet';
+import * as countLines from 'metro/src/lib/countLines';
+
+type SourceMap = Record<string, unknown>;
+
+type ExpectedSerializedConfigThisContext = Partial<SerializerConfigT>;
+
+export type VirtualJSOutput = {
+  type: string;
+  data: {
+    code: string;
+    lineCount: number;
+    map: [];
+  };
+};
+export type MetroSerializer = (
+  entryPoint: string,
+  preModules: ReadonlyArray<Module>,
+  graph: ReadOnlyGraph,
+  options: SerializerOptions & { sentryBundleCallback?: (bundle: Bundle) => Bundle },
+) => string | { code: string; map: string } | Promise<string | { code: string; map: string }>;
+
+type MetroModuleId = number;
+type MetroModuleCode = string;
+
+type Bundle = {
+  modules: [MetroModuleId, MetroModuleCode][];
+  post: string;
+  pre: string;
+};
+
+const DEBUG_ID_PLACE_HOLDER = '__debug_id_place_holder__';
+const DEBUG_ID_MODULE_PATH = '__debugid__';
+const PRELUDE_MODULE_PATH = '__prelude__';
+const SOURCE_MAP_COMMENT = '//# sourceMappingURL=';
+const DEBUG_ID_COMMENT = '//# debugId=';
+
+const createDebugIdSnippet = (debugId: string): string => {
+  return `var _sentryDebugIds={},_sentryDebugIdIdentifier="";void 0===_sentryDebugIds&&(_sentryDebugIds={});try{var stack=(new Error).stack;stack&&(_sentryDebugIds[stack]="${debugId}",_sentryDebugIdIdentifier="sentry-dbid-${debugId}")}catch(e){}`;
```

Comment:
How do we ensure this stays in sync with `scripts/sentry-debugid-injection-snippet.js`? Should we write a test?

Reply:
Added a test for this.